go/tools/builders/generate_checker_main.go

```diff
@@ -31,15 +31,206 @@ import (
+		pkgs:     pkgs,
+		stdlib:   stdlib,
+	}
+	apkg, err := load(fset, imp, flag.Args())
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "checker: error loading package: %v\n", err)
+		return
+	}
+
+	c := make(chan result)
+	for _, a := range analysis.Analyses() {
+		go func() {
+			defer func() {
+				// Prevent a panic in a single analysis from interrupting other analyses.
+				if r := recover(); r != nil {
+					c <- result{err: fmt.Errorf("recovered from analysis panic: %v", r)}
+				}
+			}()
+			res, err := a.Run(apkg)
+			if err != nil {
+				if err != analysis.ErrSkipped {
+					c <- result{err: fmt.Errorf("analysis failed: %v", err)}
+				}
+				c <- result{}
+			}
+			c <- result{findings: res.Findings}
+		}()
+	}
+	var allFindings []*analysis.Finding
+	for i := 0; i < len(analysis.Analyses()); i++ {
+		result := <-c
+		if result.err != nil {
+			fmt.Fprintf(os.Stderr, "checker: %v\n", result.err)
+			continue
+		}
+		allFindings = append(allFindings, result.findings...)
+	}
+	if len(allFindings) != 0 {
+		fmt.Fprintln(os.Stderr, "errors found during build-time code analysis:")
+		for _, f := range allFindings {
+			fmt.Fprintf(os.Stderr, "%s: %s\n", fset.Position(f.Pos), f.Message)
+		}
+		os.Exit(1)
+	}
+}
+
+type result struct{
+	findings []*analysis.Finding
+	err error
+}
+
+// load parses and type checks the source code in each file in filenames.
+// On failure, it reports errors to Error and returns (nil, nil, nil).
+func load(fset *token.FileSet, imp types.Importer, filenames []string) (*analysis.Package, error) {
+	if len(filenames) == 0 {
+		return nil, errors.New("no filenames")
+	}
+	var files []*ast.File
+	for _, file := range filenames {
+		f, err := parser.ParseFile(fset, file, nil, parser.ParseComments)
+		if err != nil {
+			return nil, err
+		}
+		files = append(files, f)
+	}
+
+	config := types.Config{
+		Importer: imp,
+		Error: func(err error) {
+			e := err.(types.Error)
+			msg := fmt.Sprintf("%s", e.Msg)
+			posn := e.Fset.Position(e.Pos)
+			if posn.Filename != "" {
+				msg = fmt.Sprintf("%s: %s", posn, msg)
+			}
+			fmt.Fprintln(os.Stderr, msg)
+		},
+	}
+	info := &types.Info{
+		Types:     make(map[ast.Expr]types.TypeAndValue),
+		Uses:      make(map[*ast.Ident]types.Object),
+		Defs:      make(map[*ast.Ident]types.Object),
+		Implicits: make(map[ast.Node]types.Object),
+	}
+	pkg, err := config.Check(files[0].Name.Name, fset, files, info)
+	if err != nil {
+		// Errors were already reported through config.Error.
+		return nil, nil
+	}
+	return &analysis.Package{Fset: fset, Files: files, Types: pkg, Info: info}, nil
+}
+
+type importer struct {
+	fset     *token.FileSet
+	packages map[string]*types.Package
+	// pkgs maps import paths to the path to the archive file representing the
+	// corresponding library.
+	pkgs map[string]string
+	// stdlib is the root directory containing standard library package archive files.
+	stdlib string
+}
+
+func (i importer) Import(path string) (*types.Package, error) {
```

Comment:
Nit: receiver should be `*importer`. It seems like this type isn't mutated here, so maybe it doesn't matter, but `importer` doesn't seem like something that should be passed by value.

Reply:
So as you said, since everything in `importer` is either a reference type or is never mutated, it doesn't actually matter in this case. However, it requires some thought to verify that invariant, and it would be easy to change this type accidentally so that invariant is no longer true.

There are some guidelines at [CodeReviewComments](https://github.com/golang/go/wiki/CodeReviewComments#receiver-type) on the golang wiki. I don't believe this falls directly under any of those cases, so "Finally, when in doubt, use a pointer receiver."